Run wp-prettier in pre-commit hook

reykjalin opened the pull request:
This PR adds https://github.com/Automattic/wp-prettier to our dev dependencies and formats our files during commit hooks. The configuration is ripped straight from wc-admin.

I thought this would be a pretty interesting experiment, and might be a nice way to keep us honest about sticking to code standards.

I needed to make a couple of changes to our dev dependencies for this to work properly. As noted in the discussions below there were some conflicts with ESLint, and the same happened with Stylelint.

In order to fix these conflicts I had to

1. Update Stylelint
1. Install some eslint and stylelint configs that disable linting rules that conflict with prettier
1. Add special rules in the eslint and stylelint configurations to fit the WP standard, as opposed to the traditional JS/CSS standards.

Because of the updated configuration for the CSS linter I actually had to make a couple of minor changes to some CSS files, but those changes can pretty much be ignored during the review.

#### Changes proposed in this Pull Request

* Run wp-prettier on commit.
* Adds CSS and JS formatting commands via wp-prettier:
    * `npm run format` - formats **all** JS and CSS files in the project.
    * `npm run format:js` - formats **all** JS files in the project.
    * `npm run format:css` - formats **all** CSS files in the project.
    * `npm run format:provided <file_1> <file_2> '<glob>'` - formats the provided files and/or globs.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Make some changes that don't follow our coding standards
* Commit changes
* Observe that the file has been formatted

OR

* Make some changes that don't follow our coding standards
* run `npm run format:provided <changed_file>`
* Observe that the file has been formatted

-------------------

- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

reykjalin commented on package.json:
```diff
@@ -28,6 +28,7 @@
     "lint:css": "stylelint 'client/**/*.scss' 'client/**/*.css' 'assets/**/*.scss' 'assets/**/*.css'",
     "lint:css-fix": "npm run lint:css -- --fix",
     "lint:js": "eslint . --ext=js,jsx",
+    "lint:js-fix": "prettier --write",
```
I've changed it to `format` 🙂.

I considered using `prettify`, but since prettier is a formatter I thought `format` would be more fitting.

reykjalin commented on package.json:
```diff
@@ -37,7 +38,7 @@
   },
   "lint-staged": {
-    "*.{js,jsx}": "npm run lint:js",
+    "*.{js,jsx}": "npm run lint:js-fix",
```
I disagree with this actually.

1. The linter is fine while you're developing, but it doesn't catch all issues. Prettier is a very opinionated formatter and makes changes that aren't required by the linter.
1. I've actually managed to fix all the conflicts in our codebase, without modifying the code (aside from a necessary ignore in a CSS file) 🎉  I tested this by formatting all the files and then running the linters 🙂. Please try this too!
1. We should use format on save, I agree. But not everyone will have that configured by default. Formatting on the pre-commit hook will automatically take care of that.

In addition to this wc-admin has also set the precedent for running prettier in a pre-commit hook.

If you have strong feelings on this I'd love to hear some more arguments for **not** running prettier in the pre-commit hook, but personally I think we **should** run prettier in the pre-commit hook. For me this basically boils down to enforcement vs. checking, and I think we should **enforce** the prettier style.

reykjalin commented on .prettierrc:
```diff
@@ -0,0 +1,10 @@
+parenSpacing: true
+jsxBracketSameLine: false
+semi: true
+arrowParens: always
```
Thanks for pointing this out! I've fixed the config file now 🙂. Don't know why it wasn't right in the first place 🤔.

>How about adding .prettierignore

I don't think this is necessary. Right now all the formatting commands only look for JS and CSS files. There's also a command to selectively format only provided files.

The pre-commit hook is also only run on files you actually changed, so I don't think this will introduce a significant enough delay in commits that it's worth it to try to speed it up?

vbelolapotkov replied:
> I don't think this is necessary. Right now all the formatting commands only look for JS and CSS files under assets and client. There's also a command to selectively format only provided files.

This should improve scanning speed for js and css files. But I agree we can add it later if necessary

reykjalin commented:
I've updated the PR with the following changes:

1. Removed the demo changes to the whole project.
1. The pre-commit hook now formats all changed files instead of just linting them.
1. Changed the name of some of the commands:
    - `npm run lint:js-fix` is now `npm run format:js` and uses `prettier`.
    - `npm run lint:css-fix` is now `npm run format:css` and uses `prettier`.
1. Added some commands:
    - `npm run format` formats **all** JS and CSS files in the project.
    - `npm run format:provided <file_1> <file_2> '<glob>'` formats the provided files and/or globs

It's ready for an actual review! 🎉

vbelolapotkov commented on package.json:
```diff
@@ -35,19 +35,21 @@
       "pre-commit": "lint-staged"
     }
   },
   "lint-staged": {
-    "*.{js,jsx}": "npm run lint:js",
-    "*.scss": "npm run lint:css",
+    "*.{js,jsx,scss,css}": "npm run format:provided",
```
I'm still concerned with this change. It's basically disables linter checks and relies on formatter only. 
Side-effect - linter scripts (lint:js, lint:css) seem to be not used anymore.

The other outcome is if you just remove a single space in the current file e.g. `client/deposits/details/index.js` and commit the change. The committed diff will be bloated with other formatting changes introduced by prettier which you didn't staged.

As I said before, this is my personal opinion on the topic and I wouldn't like forcing it. I'd ask someone else here for another opinion.

DanReyLop replied:
LOL, we reached the same conclusions almost at the same time.

> I'm still concerned with this change. It's basically disables linter checks and relies on formatter only.
Side-effect - linter scripts (lint:js, lint:css) seem to be not used anymore.

https://github.com/Automattic/woocommerce-payments/pull/631#discussion_r456360981

> The committed diff will be bloated with other formatting changes introduced by prettier which you didn't staged.

https://github.com/Automattic/woocommerce-payments/pull/631#pullrequestreview-450551039

reykjalin replied:
I think this has been resolved in https://github.com/Automattic/woocommerce-payments/pull/631#issuecomment-662374574, what do you think?

DanReyLop commented on package.json:
```diff
@@ -35,19 +35,21 @@
   "lint-staged": {
-    "*.{js,jsx}": "npm run lint:js",
-    "*.scss": "npm run lint:css",
+    "*.{js,jsx,scss,css}": "npm run format:provided",
```
`eslint` can catch some mistakes that `prettier` can't. This should be:

```suggestion
    "*.{js,jsx}": ["npm run format:provided", "eslint"],
    "*.{scss,css}": ["npm run format:provided", "stylelint"],
```

To reproduce, declare a variable in JS but never use it. That's an error that only `eslint` can catch. That should fail the pre-commit hook.

vbelolapotkov replied:
I'd suggest following changes on top of @DanReyLop 's suggestions:

* `eslint` => `npm run lint:js`
* `stylelint` => `npm run lint:css`

DanReyLop replied:
`npm run lint:js` runs the linter on all the files though, that's inefficient. Same for `npm run lint:css`. We just need to run them on the "staged" files on the pre-commit hook.

vbelolapotkov replied:
AFAIK lint-staged handles that.

vbelolapotkov replied:
Made local test and confirm it works like that:

1. Change two js files. Add unused const to one of them.
2. Stage only correct file.
3. Commit changes.

Result: commited successfully, unstaged file does not affect commit.

reykjalin replied:
Added linting in 774e74f

DanReyLop commented on assets/css/admin.css:
```diff
@@ -10,6 +10,7 @@
 	font-style: normal;
 }
 
+/* stylelint-disable selector-id-pattern */
```
The JS and CSS files in `assets` aren't generated, they're lovingly hand-crafted :)

The reasoning (at least for some of them) is that they're powering the customer-side checkout, so size if more important than in the backend. Webpack/Babel/SCSS would add some bloat.

vbelolapotkov replied:
That makes sense, thank you.

reykjalin commented on package.json:
```diff
@@ -35,19 +35,21 @@
-    "lint:php-fix": "./vendor/bin/phpbf --standard=phpcs.xml.dist $(git ls-files | grep .php$)"
+    "lint:php-fix": "./vendor/bin/phpbf --standard=phpcs.xml.dist $(git ls-files | grep .php$)",
+    "format": "npm run format:js && npm run format:css",
+    "format:js": "npm run format:provided 'assets/**/*.js' 'client/**/*.js'",
```
I used one of your other suggestions and added a `.prettierignore` in https://github.com/Automattic/woocommerce-payments/pull/631/commits/38d3c867e89a291dc5afac467b16b79f30fc4147, and instead just `npm run format:provided '**/*.js'` and `npm run format:provided '**/*.scss' '**/*.css'`.

The `.prettierignore` ignores:

```
bin
dist
docker
node_modules
release
vendor
wordpress-org-assets
```

Thanks for the suggestion here! I didn't think about the paths outside `client` and `assets` 😄

vbelolapotkov replied:
> Thanks for the suggestion here! I didn't think about the paths outside client and assets 😄

👍 

> The .prettierignore ignores:

Could you please add the following paths to .prettierignore as well:
```
tests/e2e/docker
tests/e2e/deps
```

reykjalin replied:
Added in 7196861cada99cea3271c3b939339016befa53b8 🙂

reykjalin commented:
>The other outcome is if you just remove a single space in the current file e.g. `client/deposits/details/index.js` and commit the change. The committed diff will be bloated with other formatting changes introduced by prettier which you didn't staged.

>Also, how do you think we should handle the reformatting of all files?

@DanReyLop @vbelolapotkov You're both right in bringing this up, and I think I agree with @DanReyLop here:

>Having a PR reformatting everything has its own drawbacks, but I honestly prefer it to the alternative

So I've opened #798 with only the formatting + config fixes necessary for the linters to pass CI. As a side-effect of merging that PR, the diff in this one a bit smaller and more focused.

Merging #798 will, of course, mean that other PRs will need to rebase on master and run `npm run format` (or just commit) to resolve conflicts. However, I think we generally rebase before merging PRs _anyway_ so I don't see that as a huge cost.

What do you think?
